server/src/persistence/file.rs

```diff
@@ -0,0 +1,297 @@
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot load tokens from backup, opening backup file failed".to_string(),
+                )
+            })?;
+
+        let mut contents = vec![];
+
+        file.read_to_end(&mut contents).await.map_err(|_| {
+            EdgeError::PersistenceError(
+                "Cannot load tokens from backup, reading backup file failed".to_string(),
+            )
+        })?;
+        serde_json::from_slice(&contents).map_err(|_| {
+            EdgeError::PersistenceError(
+                "Cannot load tokens from backup, parsing backup file failed".to_string(),
+            )
+        })
+    }
+
+    async fn save_tokens(&self, tokens: Vec<EdgeToken>) -> EdgeResult<()> {
+        let mut file = tokio::fs::File::create(self.token_path())
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot write tokens to backup. Opening backup file for writing failed"
+                        .to_string(),
+                )
+            })?;
+        file.write_all(
+            &serde_json::to_vec(&tokens)
+                .map_err(|_| EdgeError::PersistenceError("".to_string()))?,
+        )
+        .await
+        .map_err(|_| EdgeError::PersistenceError("Could not serialize tokens to disc".to_string()))
+        .map(|_| ())
+    }
+
+    async fn load_features(&self) -> EdgeResult<HashMap<String, ClientFeatures>> {
+        let mut file = tokio::fs::File::open(self.features_path())
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot load features from backup, opening backup file failed".to_string(),
+                )
+            })?;
+
+        let mut contents = vec![];
+
+        file.read_to_end(&mut contents).await.map_err(|_| {
+            EdgeError::PersistenceError(
+                "Cannot load features from backup, reading backup file failed".to_string(),
+            )
+        })?;
+        let contents: Vec<(String, ClientFeatures)> =
+            serde_json::from_slice(&contents).map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot load features from backup, parsing backup file failed".to_string(),
+                )
+            })?;
+        Ok(contents.into_iter().collect())
+    }
+
+    async fn save_features(&self, features: Vec<(String, ClientFeatures)>) -> EdgeResult<()> {
+        let mut file = tokio::fs::File::create(self.features_path())
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot write tokens to backup. Opening backup file for writing failed"
+                        .to_string(),
+                )
+            })?;
+        file.write_all(
+            &serde_json::to_vec(&features)
+                .map_err(|_| EdgeError::PersistenceError("".to_string()))?,
```

Comment:
Should we add an error message?

Reply:
This should never happen. _should_ never happen. This would occur as a result of the serialisation failing but that _should_ never happen since we deserialised it recently. An expect might be better here but I'm not quite confident enough to panic the thread if it does

...but also yes, I've updated the error message